oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/ImportOperationsTest.java

```diff
@@ -0,0 +1,222 @@
+        try {
+            test.setProperty("jcr:created", false);
+            session.save();
+            test.addMixin(NodeType.MIX_CREATED);
+            session.save();
+            // surprise: setting the mixin succeeds, and the subsequent state of
+            // the node is inconsistent with the mixin's type definition
+            assertEquals(false, test.getProperty("jcr:created").getBoolean());
+        } finally {
+            test.remove();
+            session.save();
+        }
+    }
+
+    @Test
+    public void jcrMixinReferenceableOnNtUnstructuredBeforeSettingMixin() throws RepositoryException {
+        Node test = testNode.addNode(TEST_NODE_NAME, NodeType.NT_UNSTRUCTURED);
+        try {
+            String testUuid = UUID.randomUUID().toString();
+            test.setProperty("jcr:uuid", testUuid);
+            session.save();
+            test.addMixin(NodeType.MIX_REFERENCEABLE);
+            session.save();
+            // surprise: setting the mixin succeeds, and the subsequent state of
+            // the node doesn't have an autocreated UUID
+            assertEquals(testUuid, test.getProperty("jcr:uuid").getString());
+        } finally {
+            test.remove();
+            session.save();
+        }
+    }
+
+    @Test
+    public void jcrMixinReferenceableOnNtUnstructuredAfterSettingMixin() throws RepositoryException {
+        Node test = testNode.addNode(TEST_NODE_NAME, NodeType.NT_UNSTRUCTURED);
+        test.addMixin(NodeType.MIX_REFERENCEABLE);
+        session.save();
+        try {
+            String testUuid = UUID.randomUUID().toString();
+            test.setProperty("jcr:uuid", testUuid);
+            session.save();
+            fail("Setting jcr:uuid after adding mixin:referenceable should fail");
+        } catch (ConstraintViolationException ex) {
+            // expected
+        } finally {
+            test.remove();
+            session.save();
+        }
+    }
+
+    @Test
+    public void jcrMixinReferenceableOnNtUnstructuredAfterRemovingMixin() throws RepositoryException {
+        Node test = testNode.addNode(TEST_NODE_NAME, NodeType.NT_UNSTRUCTURED);
+        test.addMixin(NodeType.MIX_REFERENCEABLE);
+        session.save();
+        try {
+            test.removeMixin(NodeType.MIX_REFERENCEABLE);
+            String testUuid = UUID.randomUUID().toString();
+            test.setProperty("jcr:uuid", testUuid);
+            session.save();
+            test.addMixin(NodeType.MIX_REFERENCEABLE);
+            session.save();
+            assertEquals(testUuid, test.getProperty("jcr:uuid").getString());
+            Node check = session.getNodeByIdentifier(testUuid);
+            assertTrue(test.isSame(check));
+        } finally {
+            test.remove();
+            session.save();
+        }
+    }
+
+    @Test
+    public void jcrMixinReferenceableOnNtUnstructuredAfterRemovingMixinButDanglingReference() throws RepositoryException {
```

Comment:
JCR TCK?

Reply:
Maybe; if you can find it. For now I'll leave it here, because having related tests in different places doesn't seem to be helpful.